lute/fs/src/fs_impl.cpp

```diff
@@ -0,0 +1,261 @@
+
+    // Zero out chunk buffer before next read
+    std::fill(r->chunk.begin(), r->chunk.end(), 0);
+
+    // Continue reading from current position
+    uv_fs_read(uv_default_loop(), r->get(), r->file->fd.value(), &r->iov, 1, -1, FSRead::readCallback);
+}
+
+void FSWrite::writeCallback(uv_fs_t* req)
+{
+    auto w = FSWrite::from(req);
+    auto bytesWritten = req->result;
+
+    if (bytesWritten < 0)
+    {
+        w->fail("Error writing file: %s", uv_strerror(bytesWritten));
+        delete w;
+        return;
+    }
+
+    w->offset += bytesWritten;
+    if (w->offset == w->toWrite.size())
+    {
+        w->succeed(
+            [](lua_State* L)
+            {
+                return 0;
+            }
+        );
+
+        delete w;
+        return;
+    }
+
+    // Copy next chunk to write
+    size_t remaining = w->toWrite.size() - w->offset;
+    size_t chunkSize = std::min(remaining, w->chunk.size());
+    std::copy(w->toWrite.begin() + w->offset, w->toWrite.begin() + w->offset + chunkSize, w->chunk.begin());
+    w->iov = uv_buf_init(w->chunk.data(), chunkSize);
+
+    uv_fs_write(uv_default_loop(), w->get(), w->file->fd.value(), &w->iov, 1, -1, FSWrite::writeCallback);
+}
+
+int read_impl(lua_State* L, UVFile* handle)
+{
+    if (!handle->fd.has_value())
+    {
+        luaL_errorL(L, "File handle is closed");
+    }
+
+    auto req = new FSRead(L, handle);
+    uv_fs_read(uv_default_loop(), req->get(), handle->fd.value(), &req->iov, 1, -1, FSRead::readCallback);
+
+    return lua_yield(L, 0);
+}
+
+int write_impl(lua_State* L, UVFile* handle, const char* toWrite, size_t numBytes)
+{
+    if (!handle->fd.has_value())
+    {
+        luaL_errorL(L, "File handle is closed");
+    }
+
+    auto req = new FSWrite(L, handle, toWrite, numBytes);
+
+    // Copy first chunk to write
+    size_t chunkSize = std::min(numBytes, req->chunk.size());
+    std::copy(req->toWrite.begin(), req->toWrite.begin() + chunkSize, req->chunk.begin());
+    req->iov = uv_buf_init(req->chunk.data(), chunkSize);
+
+    uv_fs_write(uv_default_loop(), req->get(), handle->fd.value(), &req->iov, 1, -1, FSWrite::writeCallback);
+    return lua_yield(L, 0);
+}
+
+int close_impl(lua_State* L, UVFile* handle)
+{
+    if (!handle->fd.has_value())
+    {
+        delete handle;
+        luaL_errorL(L, "File handle is already closed");
+    }
+
+    auto req = new FSClose(L, handle);
+    uv_fs_close(
+        uv_default_loop(),
+        req->get(),
+        handle->fd.value(),
+        [](uv_fs_t* req)
+        {
+            auto r = FSClose::from(req);
+            auto result = req->result;
+
+            if (result < 0)
+            {
+                r->fail("Error closing file: %s", uv_strerror(result));
+                delete r;
+                return;
+            }
+
+            r->succeed(
+                [](lua_State* L)
+                {
+                    return 0;
+                }
+            );
+
+            delete r;
```

Comment:
Why do we need a manual delete here? I would expect RAII to do exactly this.

Reply:
Yup - I've gone ahead and made this work via RAII + the other places.